src/rmc.rs

```diff
@@ -33,79 +33,72 @@ impl<H: Signable, S: Signature, M: PartialMultisignature> Message<H, S, M> {
     }
 }
 
+#[derive(Clone)]
 pub enum Task<H: Signable, MK: MultiKeychain> {
     BroadcastMessage(Message<H, MK::Signature, MK::PartialMultisignature>),
 }
 
 #[async_trait]
-pub trait TaskScheduler<T> {
+pub trait TaskScheduler<T>: Send {
     fn add_task(&mut self, task: T);
     async fn next_task(&mut self) -> Option<T>;
 }
 
 #[derive(Clone, Debug, Eq, PartialEq)]
 struct ScheduledTask<T> {
-    scheduled_time: time::Instant,
     task: T,
     delay: time::Duration,
 }
 
 impl<T> ScheduledTask<T> {
-    fn new(scheduled_time: time::Instant, task: T, delay: time::Duration) -> Self {
-        ScheduledTask {
-            scheduled_time,
-            task,
-            delay,
-        }
-    }
-}
-
-impl<T: Ord> Ord for ScheduledTask<T> {
-    fn cmp(&self, other: &Self) -> Ordering {
-        other
-            .scheduled_time
-            .cmp(&self.scheduled_time)
-            .then_with(|| self.task.cmp(&other.task))
-            .then_with(|| self.delay.cmp(&other.delay))
-    }
-}
-
-impl<T: Ord> PartialOrd for ScheduledTask<T> {
-    fn partial_cmp(&self, other: &Self) -> Option<Ordering> {
-        Some(self.cmp(other))
+    fn new(task: T, delay: time::Duration) -> Self {
+        ScheduledTask { task, delay }
     }
 }
 
 pub struct DoublingDelayScheduler<T> {
     initial_delay: time::Duration,
-    scheduled_tasks: BinaryHeap<ScheduledTask<T>>,
+    scheduled_instants: BinaryHeap<Reverse<time::Instant>>,
+    scheduled_tasks: HashMap<time::Instant, ScheduledTask<T>>,
     on_new_task_tx: UnboundedSender<T>,
     on_new_task_rx: UnboundedReceiver<T>,
 }
 
-impl<T: Ord> DoublingDelayScheduler<T> {
+impl<T> DoublingDelayScheduler<T> {
     pub fn new(initial_delay: time::Duration) -> Self {
         let (on_new_task_tx, on_new_task_rx) = unbounded();
         DoublingDelayScheduler {
             initial_delay,
-            scheduled_tasks: BinaryHeap::new(),
+            scheduled_instants: BinaryHeap::new(),
+            scheduled_tasks: HashMap::new(),
             on_new_task_tx,
             on_new_task_rx,
         }
     }
+
+    fn pop_scheduled_task(&mut self) -> Option<(time::Instant, ScheduledTask<T>)> {
+        let Reverse(instant) = self.scheduled_instants.pop()?;
+        let scheduled_task = self.scheduled_tasks.remove(&instant)?;
+        Some((instant, scheduled_task))
+    }
+
+    fn schedule_task(&mut self, instant: time::Instant, scheduled_task: ScheduledTask<T>) {
+        self.scheduled_instants.push(Reverse(instant));
+        self.scheduled_tasks.insert(instant, scheduled_task);
```

Comment:
In order to prevent possible instant collision, I replaced `Instant` with `(Instant, usize)`. This way each scheduled task is identifiable by its index of type `usize`. This allowed me to replace  `HashMap` with `Vec`. Once a scheduled task lands in the vector, it stays there indefinitely, except that the delay is doubled every time it is returned.